Use smart pointers to access LocalFrameLoaderClient

rwlbuis opened the pull request:
#### 95c09cf83d555b4f3723dd3f964746a9262731a3
<pre>
Use smart pointers in webaudio
<a href="https://bugs.webkit.org/show_bug.cgi?id=285592">https://bugs.webkit.org/show_bug.cgi?id=285592</a>

Reviewed by NOBODY (OOPS!).

WIP.

* Source/WTF/wtf/NativePromise.h:
* Source/WebCore/Modules/webaudio/AudioWorkletThread.cpp:
(WebCore::AudioWorkletThread::workerLoaderProxy):
* Source/WebCore/Modules/webaudio/AudioWorkletThread.h:
(WebCore::AudioWorkletThread::messagingProxy):
</pre>

rniwa commented on Source/WebCore/Modules/websockets/WebSocket.cpp:
```diff
@@ -334,7 +334,7 @@ ExceptionOr<void> WebSocket::connect(const String& url, const Vector<String>& pr
     auto reportRegistrableDomain = [domain = RegistrableDomain(m_url).isolatedCopy()](auto& context) mutable {
         if (RefPtr frame = downcast<Document>(context).frame())
-            frame->loader().client().didLoadFromRegistrableDomain(WTFMove(domain));
+            frame->checkedLoader()->client().didLoadFromRegistrableDomain(WTFMove(domain));
```
This is odd. `client()` is clearly a trivial function. We shouldn't need to use `checkedLoader` here.

rwlbuis replied:
client() is trivial but didLoadFromRegistrableDomain is not. A similar change was done in 273326@main. Shouldn't transitivity apply here?

rniwa replied:
No. If didLoadFromRegistrableDomain is non-trivial, then client() is the one that needs to return Ref/RefPtr variant.

rwlbuis replied:
The client is stored using a unique_ptr in FrameLoader, so I feel it is basically part of FrameLoader and FrameLoader still needs to be checked on these operations. @cdumez  WDYT? I guess we should either adopt the new changes or undo the existing ones, I am fine with doing it either way, just let me know.

rniwa replied:
A better approach is for FrameLoaderClient to have ref/deref functions that forwards ref/deref to FrameLoader.

rwlbuis replied:
I see what you mean. I started implementing that, however it is cumbersome given the FrameLoaderClient inheritance structure and the number of call sites to update. What do you think about my latest revision that follows the BackForwardController solution?

rwlbuis committed:
Use smart pointers to access LocalFrameLoaderClient
https://bugs.webkit.org/show_bug.cgi?id=277071

Reviewed by NOBODY (OOPS!).

Make the LocalFrameLoaderClient itself be ref counted
by FrameLoader so no special smart pointers have to
be used to access it in typical client() calls.

* Source/WebCore/loader/EmptyClients.cpp:
(WebCore::pageConfigurationWithEmptyClients):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::FrameLoader):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/LocalFrameLoaderClient.h:
* Source/WebCore/page/LocalFrame.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/Storage/WebSharedWorkerContextManagerConnection.cpp:
(WebKit::WebSharedWorkerContextManagerConnection::launchSharedWorker):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::createSubframe):
(WebKit::WebFrame::createProvisionalFrame):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::mainFrameCreationParameters):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(+[WebFrame _createFrameWithPage:frameName:frameView:ownerElement:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView _commonInitializationWithFrameName:groupName:]):
(-[WebView initSimpleHTMLDocumentWithStyle:frame:preferences:groupName:]):

cdumez commented on Source/WebCore/loader/LocalFrameLoaderClient.h:
```diff
@@ -103,7 +103,7 @@ enum class WasPrivateRelayed : bool;
 struct StringWithDirection;
 
-class WEBCORE_EXPORT LocalFrameLoaderClient : public FrameLoaderClient {
+class WEBCORE_EXPORT LocalFrameLoaderClient : public FrameLoaderClient, public RefCounted<LocalFrameLoaderClient> {
```
My alternative proposal:
1. Don't subclass RefCounted here
2. Have LocalFrameLoaderClient forward its ref-counting to FrameLoader
3. Have FrameLoader forward its ref counting to LocalFrame

cdumez replied:
@rniwa @geoffreygaren What do you think?

rniwa replied:
So FrameLoader isn't ref counted either so it seems that we need to make it also forward ref to LocalFrame?

cdumez replied:
Yes, this is step 3 of my plan above.

rwlbuis replied:
I don't see how step 2 is possible. Anyway, I think it is better if somebody else works on this, since I likely won't be able to work on this in October.

cdumez replied:
Starting: https://github.com/WebKit/WebKit/pull/34656

cdumez replied:
Alternative in https://github.com/WebKit/WebKit/pull/34705

rwlbuis replied:
Closing this now https://github.com/WebKit/WebKit/pull/34705 landed.
